parameter setup in v3 API

bertsky opened the pull request:
addresses [the problem with required parameters](https://github.com/OCR-D/core/pull/1240#issuecomment-2304121310)

@kba @MehmedGIT could you please review?

kba committed:
:package: v3.0.0a2

kba committed:
remove make *-workaround, we will not do that for v3+

Commit:
Processor.parameter: only validate when set…

- `Processor` init: do not validate, only set `parameter` if present
  (needed to avoid exception from `ParameterValidator` for processors
   with mandatory params in non-processing usage)
- `Processor.parameter`: allow `None`, but validate when set
- `Processor._setup`: validate parameters, then call `Processor.setup`

Commit:
get_processor: ensure passing non-empty parameter, rely on `_setup` to call `setup`

Commit:
test_processor: adapt, check required parameters

kba commented on src/ocrd/processor/base.py:
```diff
@@ -289,6 +295,17 @@ def list_resources(self):
+    def _setup(self) -> None:
+        """
+        Validate parameters, then run :py:meth:`setup`. Called whenever
+        :py:data:`parameter` changes.
```
```suggestion
        :py:data:`parameter` is re-assigned.
```

Because "changes" implies that `_setup` is called when doing e.g. `self.parameter['foo'] = 'bar'` when it is only triggered by the setter, e.g. `self.parameter = {'foo': 'bar'}`.

bertsky replied:
Oh, indeed, that's very significant!

So perhaps in addition to that we should indeed use [collections.Mapping](https://docs.python.org/2/library/collections.html#collections.Mapping) instead of dict here, as you proposed?

kba commented on src/ocrd/processor/helpers.py:
```diff
@@ -374,28 +374,29 @@ def get_cached_processor(parameter: dict, processor_class):
 def get_processor(
         processor_class,
-        parameter: dict,
+        parameter: Optional[dict],
```
Any reason to still have this as a required arg?
```suggestion
        parameter: Optional[dict] = None,
```

bertsky replied:
Not really, no. But then `get_cached_processor` should be changed analogously.

kba commented on tests/processor/test_processor.py:
```diff
@@ -95,8 +96,18 @@ def test_json(self):
+        with self.assertRaisesRegex(ValueError, 'is a required property'):
+            proc.parameter = {}
+        with self.assertRaisesRegex(ValueError, 'is a required property'):
+            get_processor(DummyProcessorWithRequiredParameters, None)
```
```suggestion
            get_processor(DummyProcessorWithRequiredParameters)
```

kba commented on src/ocrd/processor/base.py:
```diff
@@ -289,6 +295,17 @@ def list_resources(self):
+        parameterValidator = ParameterValidator(self.ocrd_tool)
+        report = parameterValidator.validate(self.parameter)
+        if not report.is_valid:
+            raise ValueError("Invalid parameters %s" % report.errors)
```
```suggestion
            raise ValueError(f'Invalid parameters:\n{report.to_xml()}')
```

bertsky commented:
Unfortunately, by always defining a `Processor.parameter` regardless of processing or non-processing use, we break implementations like ocrd_tesserocr:
https://github.com/OCR-D/ocrd_tesserocr/blob/dcbd5227f834c219e36d5dc41606a08e85e67a15/ocrd_tesserocr/segment_region.py#L16

We had the same with `Processor.output_file_grp` checks, which led me to change v3 so this would not immediately break such processors. In this case, I am sure it's only ocrd_tesserocr, which we will migrate immediately, anyway. (But will all installations which update core to v3 then also update ocrd_tesserocr?)

I could still change this PR to not provide `Processor.parameter == None` but rather `not hasattr(Processor, 'parameter')` right until setting parameters explicitly (in get_processor or elsewhere)...

kba reviewed (approved):
The parameter validation happening on assignment is much better and further reduces the complexity of the constructor. And with the `_setup -> validation -> setup` mechanism we're certain that everything is properly (re-)loaded when the parameters change - which I still don't think should happen, but if it is being done, we're safe.

Also just tested it by making `model` of `ocrd-kraken-recognize` required and `--help` works again.

So, LGTM!

bertsky committed:
improve _setup docstring

Co-authored-by: Konstantin Baierer <kba@users.noreply.github.com>

bertsky committed:
Processor._setup: raise with full ParameterValidator report

Co-authored-by: Konstantin Baierer <kba@users.noreply.github.com>

kba commented on src/ocrd/processor/base.py:
```diff
@@ -139,6 +139,19 @@ def ocrd_tool(self) -> dict:
+
+    @parameter.setter
+    def parameter(self, parameter : dict) -> None:
+        self._parameter = parameter
```
```suggestion
        from types import MappingProxyType
        self._parameter = MappingProxyType(parameter)
```

not tested but this should essentially make `self.parameter` read-only but still allow assignment. So processor developers could still do

```
self.parameter = {**self.parameter, 'foo': 'bar'}
```

if they need it, with the benefit of `_setup`.

There are ways to make that transparent but since I don't think hard-coding parameter assignments should be done, I think it's reasonable to make it a bit more complicated.

bertsky replied:
Wow, that looks very promising. I'll try it out and add some tests.

bertsky replied:
Unfortunately, that conflicts with our default expansion in the ParameterValidator.

bertsky replied:
It also begs the question whether we actually need `_setup` as an intermediary. We can simply do the parameter validation in the parameter setter, and then call `setup` directly.

bertsky replied:
Also, when we do `json.dumps(self.parameter)`, we end up with
```
TypeError: Object of type mappingproxy is not JSON serializable
```

bertsky replied:
Fixed in f2d53a6.

However, even if I cannot effectively change the parameters, I also get no exception or warning about it.

bertsky replied:
So perhaps a frozendict would be better?

bertsky replied:
Oh, that may be just because I always return a `dict(.)` copy of `_parameter` already. Which, thinking of it, in itself would be enough to prevent overwriting the contents. But I do want to get an exception, really...

bertsky replied:
Ah, but then again, with `frozendict` we would not need the `dict(.)` trick, because it is directly serializable: 7297ca2d

kba replied:
> Oh, that may be just because I always return a `dict(.)` copy of `_parameter` already.

If it's to prevent modifying the original dict, doing `json.loads(json.dumps(...))` is the fastest way to do a deep clone. But why is that necessary here?

> So perhaps a frozendict would be better?

Oh right, yes, that would probably be the easiest solution and we already require that ^^

kba replied:
> It also begs the question whether we actually need `_setup` as an intermediary. We can simply do the parameter validation in the parameter setter, and then call `setup` directly.

We likely don't and it is cleaner with the parameter validation directly in the setter.

bertsky committed:
get_processor: parameter only as kwarg

Co-authored-by: Konstantin Baierer <kba@users.noreply.github.com>

Commit:
tests: adapt for get_processor parameter only as kwarg

bertsky committed:
Processor.parameter: make the bound dict read-only

Co-authored-by: Konstantin Baierer <kba@users.noreply.github.com>

Commit:
Processor.parameter: move ParameterValidator back to setter, convert to plain dict in getter for serialization etc

Commit:
Processor.parameter: frozendict instead of mappingproxy, add test

MehmedGIT reviewed (approved):
I have not tested if it works, just peeked over the code. Seems okay, I am unsure how that affects the processor caching. How about caching only the models used by the processors and not the entire processor class?

bertsky commented:
> I am unsure how that affects the processor caching. 

I myself am just trying to understand how `lru_cache` actually works in this case. It is known to require hashable types for the function arguments, but `get_cached_processor` takes a `dict` for the parameters, not a frozendict. Where do we test this?

> How about caching only the models used by the processors and not the entire processor class?

The problem is that we don't know in core where implementors store anything. We can only observe what goes in (i.e. parameters) and out (i.e. instances with bindings to large memory objects possibly even on GPU).

MehmedGIT commented:
> Where do we test this?

We do not have separate tests for that. I never tested that part to see what performance we gain from the processor caching.

Commit:
introduce Processor.shutdown to be overridden (called at deinit or parameter re-assignment)

bertsky commented:
> but `get_cached_processor` takes a `dict` for the parameters, not a frozendict

I was wrong: via `freeze_args`, it already converts dict to frozendict

BTW, I don't like the fact that `config.OCRD_MAX_PROCESSOR_CACHE` gets read on module initialization level already. That means I have no chance as a processor implementor to customize this based on what I know about my processor (like models are too large for more than 2 instances). What I would like to have is:

```python
...
from ocrd_utils import config
config.OCRD_MAX_PROCESSOR_CACHE = 2
...
from ocrd import Processor
...

class MyProcessor(Processor):
    ...
    # instances of this class will have an lru_cache size of 2 in get_processor
```

> We do not have separate tests for that. 

In 6cd4a34c I know added a simple test for `get_processor(... instance_caching=True)`.

This also contains the changes I deemed necessary for [`Processor.shutdown`](https://github.com/OCR-D/core/pull/1240#issuecomment-2304431718) – please review!

Unfortunately, I cannot properly test the combination of both: doing something in `shutdown` and checking this was in fact triggered upon uncacheing with an assertion. You can see that I commented that assertion out:
https://github.com/bertsky/core/blob/6cd4a34c689f56d4cbe5600d85978567d7b1e60e/tests/processor/test_processor.py#L123

The problem is: I can see those exact events on the console (i.e. I know that the instance was `del`ed and its `shutdown` called), but somehow, because of the weakref.finalize mechanism, I cannot capture that with pytest.

> I never tested that part to see what performance we gain from the processor caching.

It cannot trivially be done in a unit test, because there is not going to be any benefit with dummy processors. We need some concrete processor with large (memory-wise) models that take some time to load/setup. And we'd have to run a bunch of them in different parameter/model settings...

MehmedGIT commented:
> BTW, I don't like the fact that config.OCRD_MAX_PROCESSOR_CACHE gets read on module initialization level already. That means I have no chance as a processor implementor to customize this based on what I know about my processor

Right, but that cache was supposed to be used only by the `ocrd_network` module, and as a processor implementor, you were not supposed to care about that. By default, the max possible amount supported by Python should be used. The provided environment variable was introduced to provide flexibility. On the other hand, I see why keeping 128 instances is problematic. If that many instances are cached with the same big model but just different other small parameter changes - that would lead to memory peaks. Very bad in certain cases with big models! However, in a fewer cached instances, there will be more module loads/unloads happening.

The 2 main issues I see are: 
- we cannot cache only the model parameter but need to cache the entire parameter dictionary.
- there is no good way to generalize how many instances are to be cached from the core side for each specific processor.

What do you think about the following suggestion: Instead of keeping a single cache in the core for all processors, we could keep a list of caches. Each unique processor will have its separate cache. Then for each cache, the core will invoke a method to decide what is the desired max instances number. I am aware, that this still does not fully solve all the issues, but at least we get more flexibility and the implementers could override the default method to return the best number for each specific processor. 

EDIT: The `get_cached_processor` method in the `ocrd_network` context gets invoked only by the Processing Worker. Since each worker is tied up to a specific processor, each worker has a separate cache. So, it is still perfectly valid to set `OCRD_MAX_PROCESSOR_CACHE` separately for each worker in their environment. However, the limitation here is that the workers should run in different environments or on different hosts. When running on the same host, the memory consumption will potentially explode.

> It cannot trivially be done in a unit test, because there is not going to be any benefit with dummy processors.

Exactly. I rather meant I have not tested it manually with the `ocrd_network` module by utilizing different processors/models and comparing the performance with and without the processor caching with different amounts of processing worker instances.

> The problem is: I can see those exact events on the console (i.e. I know that the instance was `del`ed and its `shutdown` called), but somehow, because of the weakref.finalize mechanism, I cannot capture that with pytest.

I have no idea other than suggesting to write to a test file instead of stdout/stderr and read from the test file to assert.

MehmedGIT reviewed (approved):
Good for now to move forward.

bertsky commented:
> What do you think about the following suggestion: Instead of keeping a single cache in the core for all processors, we could keep a list of caches. Each unique processor will have its separate cache. Then for each cache, the core will invoke a method to decide what is the desired max instances number. I am aware, that this still does not fully solve all the issues, but at least we get more flexibility and the implementers could override the default method to return the best number for each specific processor.
> 
> EDIT: The `get_cached_processor` method in the `ocrd_network` context gets invoked only by the Processing Worker. Since each worker is tied up to a specific processor, each worker has a separate cache. So, it is still perfectly valid to set `OCRD_MAX_PROCESSOR_CACHE` separately for each worker in their environment. However, the limitation here is that the workers should run in different environments or on different hosts. When running on the same host, the memory consumption will potentially explode.

Yes, I know it's a separate cache for each worker. But the module level settings makes it impossible to set this **programmatically**. Of course, once I have a docker-compose.yml I can configure such environment variables. But if we will generate each docker-compose.yml in ocrd_all, then we would need to devise some customisation/inclusion mechanism first.

Couldn't we use an inline function for `get_cached_processor`, so we can read the state of `config.OCRD_MAX_PROCESSOR_CACHE` when instantiating the server (so after module initialisation)?

> I have no idea other than suggesting to write to a test file instead of stdout/stderr and read from the test file to assert.

Just tried that. Unfortunately, this does not work either. The `shutdown` (triggered by finalizer) gets called no earlier than when pytest leaves the function, so the file will only be filled after we can test it.

bertsky commented:
> Couldn't we use an inline function for `get_cached_processor`, so we can read the state of `config.OCRD_MAX_PROCESSOR_CACHE` when instantiating the server (so after module initialisation)?

Oh, that cannot work, of course – it would be in a new cache with every `get_processor` call.

Perhaps my above example,
```python
...
from ocrd_utils import config
config.OCRD_MAX_PROCESSOR_CACHE = min(config.OCRD_MAX_PROCESSOR_CACHE, 4)
...
# stuff for my processor
```
is not all wrong: it would just have to be done before importing the processor decorator, right?

MehmedGIT commented:
> But the module level settings makes it impossible to set this programmatically. 

My idea was in fact to be able to set that programmatically. Maybe I should have gone into more detail. What I am suggesting is getting rid of the `lru_cache` decorator on `get_cached_processor` and writing our wrapper for `functools.lru_cache(maxsize=128, typed=False)`. The max size to be passed will then be decided at run-time based on the `ProcessorClass.get_max_cache_size()`. By default, there will be a `get_max_cache_size()` method in core returning 128 or `config.OCRD_MAX_PROCESSOR_CACHE`. That method can then be overridden on the Processor level to return the "ideal" value. However, to prevent potential issues, we would probably need a dictionary/list of caches in the core and not a single cache. One cache for each unique processor.

Example:
```python
{
"ocrd-cis-ocropy-recognize": CacheInstance,
"ocrd-cis-ocropy-binarize": CacheInstance,
"ocrd-kraken-recognize": CacheInstance,
...
}
```

Each CacheInstance max size will depend on the `ProcessorClass.get_max_cache_size()`. Keep in mind that my suggestion is only an idea and I have not implemented something like that before. Not sure how much flexibility we can have with `functools.lru_cache`

MehmedGIT commented:
> Oh, that cannot work, of course – it would be in a new cache with every get_processor call.

Yes.

> is not all wrong: it would just have to be done before importing the processor decorator, right?

I think as long as we use the `functools.lru_cache` as a decorator for `get_cached_processor`, we will face all kinds of limitations. A different strategy is needed.

bertsky commented:
> getting rid of the `lru_cache` decorator on `get_cached_processor` and writing our wrapper for `functools.lru_cache(maxsize=128, typed=False)`. The max size to be passed will then be decided at run-time based on the `ProcessorClass.get_max_cache_size()`.

okay, but I still don't understand **where** we should wrap then. The storage can only be on a module, in a module global, or in a class. I fail to see any class we could use for that (and wrapping `lru_cache` in instances takes extra steps due to GC interference). I managed to get it working with a module-level global, but that's ugly...

Commit:
Processor: introduce `max_instances` class attribute

MehmedGIT commented:
> okay, but I still don't understand where we should wrap then. The storage can only be on a module, in a module global, or in a class. I fail to see any class we could use for that (and wrapping lru_cache in instances takes extra steps due to GC interference). I managed to get it working with a module-level global, but that's ugly...

I also do not know. I took over the cache implementation as it is from https://github.com/OCR-D/core/pull/884. If the current cache implementation is painful to adapt to v3 and requires globals and various hacks, we could also consider completely removing it.

kba committed:
Processor.metadata_location property to specify where in the package ocrd-tool.json is found

Commit:
get_cached_processor: set lru_cache maxsize from min(cfg,class) at runtime

Commit:
test get_processor instance_caching w/ max_instances

bertsky commented:
@MehmedGIT well, it is a valuable functionality as it is, and for me it's not so much a matter of adapting to v3 but taking the chance to improve it a little. See last 3 commits for said attempt via hacking global module storage. Sorry – I don't know any better.

MehmedGIT commented:
> @MehmedGIT well, it is a valuable functionality as it is, and for me it's not so much a matter of adapting to v3 but taking the chance to improve it a little. See last 3 commits for said attempt via hacking global module storage. Sorry – I don't know any better.

Thanks. I will have a look and test with the `ocrd_network` module. Please do not hinder the development by waiting for me. Merge this PR if required. I could adapt afterwards if I find a better solution.

Commit:
test get_processor instance_caching w/ clear_cache

Commit:
Processor.verify: always check cardinality (as we now have the defaults from ocrd-tool.json)

Commit:
fix --log-filename (6fc606027a): apply in ocrd_cli_wrap_processor

Commit:
fix exception

Commit:
adapt to PIL.Image moved constants

Commit:
ocrd_utils: add parse_json_file_with_comments

Commit:
cli.workspace: pass fileGrp as well, improve description

Commit:
OcrdMets.add_agent: does not have positional args

Commit:
remove misplaced kwargs from run_processor

Commit:
Processor.metadata: refactor…

- `metadata`, `executable`, `ocrd_tool`, `version`: use cached_property
  instead of internal ad-hoc attributes
- rename `metadata_location` → `metadata_filename`, add cached_property
  chain `metadata_location`, `metadata_rawdict` used by `metadata` to
  make it easy to override
  - `metadata_filename` if just the path of `ocrd-tool.json` in the package deviates
  - `metadata_location` if the `ocrd-tool.json` is not distributed via Python pkg
  - `metadata_rawdict` if the `ocrd-tool.json` is not in a file
  - `metadata` if the validated, expanded `ocrd-tool.json` is somewhere else
- `DummyProcessor`: just override `Processor.metadata_filename`
- processor tests: adapt to new properties and `verify` enforcing cardinality

Commit:
bashlib input-files: adapt, allow passing ocrd-tool.json path and executable name

Commit:
add to pylint karma

Commit:
update pylintrc

bertsky commented:
I did some more refactoring and cleanup.

Notably, because we always default-expand the schema, we can now enforce the fileGrp cardinalities already – see https://github.com/bertsky/core/pull/14/commits/5819c8167d1a2be662dd32b58bff0531ded40f8b

Also, I cherry-picked from #16 and exported the full chain of `metadata` loading, so it can be overridden at exactly where implementors like – see https://github.com/bertsky/core/pull/14/commits/7966057f975a76db39bf10b1c38540860b7c179c

I also made use of the fact that in bashlib's `ocrd__wrap` we already know where the ocrd-tool.json file is and what the executable is called: by passing this info into `ocrd bashlib input-files`, we can even better use the standard processor class and decorator.

Since no-one commented on [the problem with breaking `Processor.parameter` in v2](https://github.com/bertsky/core/pull/14#issuecomment-2304459818), I consider this irrelevant and will keep it. (So just ocrd_tesserocr must be updated+installed immediately.)

This is ready AFAICS. Does anyone have an opinion when this can/should be merged, so we can make another prerelease?

kba commented on src/ocrd/cli/bashlib.py:
```diff
@@ -129,4 +125,31 @@ def process_page_file(self, *input_files):
             print(f"[outputFileId]='{output_file_id}'")
-    ocrd_cli_wrap_processor(BashlibProcessor, **kwargs)
+    if ocrd_tool and executable:
+        class FullBashlibProcessor(BashlibProcessor):
```
clever

kba committed:
processor.metadata_location: use self.__module__ not __package__

kba commented:
> I did some more refactoring and cleanup.
> 
> Notably, because we always default-expand the schema, we can now enforce the fileGrp cardinalities already – see [5819c81](https://github.com/bertsky/core/commit/5819c8167d1a2be662dd32b58bff0531ded40f8b)

:+1:

> Also, I cherry-picked from #16 and exported the full chain of `metadata` loading, so it can be overridden at exactly where implementors like – see [7966057](https://github.com/bertsky/core/commit/7966057f975a76db39bf10b1c38540860b7c179c)

Very thorough but I needed #17 to get eynollah to work

> 
> I also made use of the fact that in bashlib's `ocrd__wrap` we already know where the ocrd-tool.json file is and what the executable is called: by passing this info into `ocrd bashlib input-files`, we can even better use the standard processor class and decorator.

:+1:


> Since no-one commented on [the problem with breaking `Processor.parameter` in v2](https://github.com/bertsky/core/pull/14#issuecomment-2304459818), I consider this irrelevant and will keep it. (So just ocrd_tesserocr must be updated+installed immediately.)

Yes, if there is breakage we can fix it quickly.
 
> This is ready AFAICS. Does anyone have an opinion when this can/should be merged, so we can make another prerelease?

I could make a `b1` right away but would like to have #17 in there, since I'm working on eynollah now which needs it.

bertsky committed:
Merge pull request #17 from OCR-D/new-processor-api-parameter-setup

processor.metadata_location: use self.__module__ not __package__

Commit:
pylint: try ignoring generateds (again)
